### DISCUSSION TITLE
Where is A1111 "entrypoint" for `controlnet.py`?

### DISCUSSION THREAD

**Discussion by holmbuar** [Mikubill/sd-webui-controlnet#1331] [Q&A] [answered]
I've been debugging the A1111 app with the ControlNet extension, in VS Code, as a Python module, with configuration
```
{
    "name": "Python: SD Module",
    "type": "python",
    "request": "launch",
    "module": "launch",
    "justMyCode": true,
    "args": ["--api", "--xformers", "--nowebui"]
}
```
in `launch.json`. 

For triggering inference, I call the A1111 API using the test example from this repo. The debugging works just fine.

I would like to create a custom `controlnet.py` file in the extensions folder, to experiment with running the extension from API only.

However, I have spent hours trying to find the "entrypoint" to the ControlNet extension, within the A1111 application. For example, in line 145 of the `extensions.py` file, I can debug and see all the extensions being initialized from the `Extension` class. The one thing I cannot see, is where the A1111 app actually calls the relevant ControlNet script file, based on the `alwayson_script` args.

Would anyone kindly nudge me in the relevant direction?

**Comment by ljleb** [comment, marked as answer]
The webui loads extension files that contain a script [here](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/89f9faa63388756314e8a1d96cf86bf5e0663045/modules/scripts.py#L257)

You may be interested in the webui's documentation for extensions if you have not read it yet: https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Developing-extensions

**Reply by holmbuar** [reply]
Thanks!

From the diagram in the documentation you linked to, I followed the breadcrumbs:

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/17097262/c64b5672-61e5-4e4f-91de-bcbbe1767016)

In addition to the script loading you refer to, I found that the `process_images_inner` method calls the `process` method

[processing.py](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/89f9faa63388756314e8a1d96cf86bf5e0663045/modules/processing.py#L585)

in

[scripts.py](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/89f9faa63388756314e8a1d96cf86bf5e0663045/modules/scripts.py#L415)

which loops through the `alwayson` scripts and calls their respective `process()` methods.
